### ISSUE TITLE
Retry function

### ISSUE THREAD

**Issue by mitsuo0114** [yokawasa/azure-log-analytics-data-collector#10]
hi, thank you for creating & maintaining useful Plugin! 
We are using this product via LogStash plugin.
https://github.com/yokawasa/azure-log-analytics-data-collector

I believe this is not LogStash Plugins issue but this repository issue.
so let me raise retry issue here. Please correct me and this issue if it is wrong.

https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api
According to manual, they require client to retry when we  following error codes.

|Code | Status | Error code | Description
-- | -- | -- | --
|429 | Too Many Requests |   | The service is experiencing a high volume of data from your account. Please retry the request later.
|500 | Internal Server Error | UnspecifiedError | The service encountered an internal error. Please retry the request.
|503 | Service Unavailable | ServiceUnavailable | The service currently is unavailable to receive requests. Please retry your request.

But currently, there is no check and because of rest-client specification, a specific exception class will be thrown when we receive these error code.
https://github.com/yokawasa/azure-log-analytics-data-collector/blob/master/lib/azure/loganalytics/datacollectorapi/client.rb#L43

https://rubydoc.info/gems/rest-client/#Exceptions__see_http___www_w3_org_Protocols_rfc2616_rfc2616-sec10_html_

but as we can imagine easily, retry function may cause not good effect like consuming memory, stacking thread.. etc and that should be configured for each purpose.
so how you think we should implement retry function for this HTTP Data collector?

**Comment by slagelwa** [comment]
Take a look at google.api_core, it has a retry decorator/method that you might be able to use.  I defined a few exceptions that can be retried and included them in the retry decorator.

**Comment by yokawasa** [comment]
@mitsuo0114 
Thank you so much for the post, and I'm ver very sorry for the late reply. I completely overlooked your post ^^; 
I think we should implement retry functions as recommended in [the API page](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api). 
If you've already any idea or implementaion that can achieve 'retry', please send PR. I'm happy to take it to the plugin.

@slagelwa Thank you for the comment. 

For the implementation, I'm simply thinking to use `ensure` keyword to implement retry like this: 
- https://blog.appsignal.com/2018/05/16/ensure-retry-and-reraise-exceptions-in-ruby.html

**Comment by yokawasa** [comment]
I'm woking on this, and will be back shortly

**Comment by yokawasa** [comment]
@mitsuo0114 

I've published azure-loganalytics-datacollector-api-0.4.0 where I added retries like this:

```ruby
          retries = 0 
          begin
            res = RestClient.post( uri, body, headers)
            res
          rescue => e
            c = e.response.code.to_i
            if c == 429 || c == 500 || c==503
              if retries < @max_retries
                retries += 1
                sleep(@retry_sleep_period)
                retry 
              end
            end
            raise e
          en
```
- https://rubygems.org/gems/azure-loganalytics-datacollector-api/versions/0.4.0
- https://rubygems.org/gems/logstash-output-azure_loganalytics/versions/0.4.0

Please use it and let me know if you find any problems.
Thank you for the issue
